### ISSUE TITLE
IndexError: list index out of range

### ISSUE THREAD

**Issue by fubar-1** [modbog/mach3linearity#1]
Hey modbog,

Thanks for sharing this.  You're welcome to just close this bug, it's mostly FYI.

I hit an 'index out of range' error in your python script until a fake 24000,24000 line was appended to the end, which wasn't a measured value.  Also, the need for the 0,0 value wasn't immediately obvious.

I'm thinking it might be a good idea to either prime the input pipeline with the first & last lines or amend your docs a bit to explain that.

Here's my `observed.csv` including the fix.  Am I doing it right?
```
0,0
1000,946
2000,1921
3000,2876
4000,3836
5000,4775
6000,5755
7000,6692
8000,7653
9000,8612
10000,9570
11000,10529
12000,11488
13000,12470
14000,13430
15000,14385
16000,15342
17000,16300
18000,17256
19000,18193
20000,19168
21000,20124
22000,21081
23000,22033
24000,22991
24000,24000
```

**Comment by modbog** [comment]
Hey Dave,

That looks correct. Based on that input you should get a .dat file that includes a list of ratios similar to:

```
ratio: 0.01
ratio: 0.010570824524312896
ratio: 0.021141649048625793
ratio: 0.03171247357293869
ratio: 0.04222365203963454
ratio: 0.05250897435897436
ratio: 0.06276538461538461
ratio: 0.07302179487179487
ratio: 0.08332942708333332
ratio: 0.09376570680628271
ratio: 0.1042369109947644
ratio: 0.11470811518324607
ratio: 0.1251624266144814
ratio: 0.13558854166666667
ratio: 0.14600520833333333
ratio: 0.15642187500000002
ratio: 0.1669141653501843
ratio: 0.17752928647497337
ratio: 0.18817891373801918
ratio: 0.19882854100106498
ratio: 0.2093254747871644
ratio: 0.21954719387755106
ratio: 0.2297512755102041
ratio: 0.23995535714285715
ratio: 0.2502814834578442
ratio: 0.2609538420490929
ratio: 0.2716262006403415
ratio: 0.28229122340425533
ratio: 0.2928655046826223
ratio: 0.30327133194588973
ratio: 0.3136771592091571
ratio: 0.32408298647242456
ratio: 0.33449817518248176
ratio: 0.34492570385818566
ratio: 0.3553532325338895
ratio: 0.36578076120959335
ratio: 0.37621395955642534
ratio: 0.38665187891440506
ratio: 0.3970902922755741
ratio: 0.40752870563674326
ratio: 0.4179608865710561
ratio: 0.42838894681960377
ratio: 0.4388164754953076
ratio: 0.44924400417101146
ratio: 0.4596715328467153
ratio: 0.4700990615224192
ratio: 0.48052659019812305
ratio: 0.49094281298299847
ratio: 0.5012321792260692
ratio: 0.5114154786150713
ratio: 0.5215987780040733
ratio: 0.5317971758664956
ratio: 0.5420963541666667
ratio: 0.5525130208333333
ratio: 0.5629296875
ratio: 0.5733507170795307
ratio: 0.5837958115183246
ratio: 0.5942670157068063
ratio: 0.604738219895288
ratio: 0.6152080607170898
ratio: 0.6256661442006269
ratio: 0.6361154649947753
ratio: 0.6465647857889237
ratio: 0.6570141065830721
ratio: 0.6674556367432151
ratio: 0.6778940501043842
ratio: 0.6883324634655531
ratio: 0.6987708768267222
ratio: 0.7092261437908495
ratio: 0.7196861924686191
ratio: 0.7301464435146443
ratio: 0.7406158747013536
ratio: 0.7512406616862325
ratio: 0.7619130202774811
ratio: 0.7725853788687299
ratio: 0.7831772053083528
ratio: 0.7935333333333331
ratio: 0.8037897435897433
ratio: 0.8140461538461536
ratio: 0.8243448990160536
ratio: 0.8347489539748952
ratio: 0.8452092050209202
ratio: 0.8556694560669453
ratio: 0.8661273036204417
ratio: 0.8765791536050155
ratio: 0.8870284743991639
ratio: 0.8974777951933123
ratio: 0.9079374836515823
ratio: 0.9184335609243697
ratio: 0.9289377626050418
ratio: 0.9394419642857141
ratio: 0.9499383363946469
ratio: 0.960377087682672
ratio: 0.9708155010438411
ratio: 0.9802346134127585
ratio: 0.9821876907302068
ratio: 0.9841407680476552
ratio: 0.9860938453651035
ratio: 0.9880469226825518
ratio: 0.99
ratio: 1.0
```

The code is not terribly robust and the interpolation bit could stand to see some improvement. It was a bit of a rush job I'm afraid. 

I will update the readme to include a note about this issue for now and hopefully spare someone else the headache of debugging.

Thanks!

**Comment by fubar-1** [comment]
I hate to trouble you but I'm still in debug mode here.  If you have a few minutes to look this over, here's what I've found so far:

- The generated Linearity.dat file doesn't seem to have any effect in spindle speed on my machine.  Mach3 doesn't seem to error (nothing in lasterror.txt, etc), and it's not trying to create a new Linearity.dat file, so I assume it's reading & silently rejecting the generated file for some reason.
- Thanks for posting your ratio: output from my observed.csv file.  I was able to reproduce the same results by tweaking the code a bit.  So I'm unsure why the Linearity.dat doesn't work.
- Hex dumping the original (default) file and the generated file, they're definitely different but I haven't gone as far as trying to decode the contents yet.

I'm attaching my resulting Linearity.dat file here.  Not begging for help (yet) but if you have any ideas...
[Linearity.zip](https://github.com/modbog/mach3linearity/files/6094144/Linearity.zip)

**Comment by modbog** [comment]
Actually, looking at the numbers again the output doesn't look correct. since your later observations show slower speeds than the requested inputs the ratios near the end should probably be positive. 

`24000/22991=1.04388673829` 

I say likely because I didn't find any documentation on how Linearity is supposed to be constructed. Just did some reverse engineering from what was there on my machine.

I've attached a Linearity.dat that might be a step in the right direction, although probably not 100% either. You can unpack it with this: https://github.com/modbog/mach3linearity/blob/main/unpack_linearity.py

[Linearity.zip](https://github.com/modbog/mach3linearity/files/6094721/Linearity.zip)

**Comment by fubar-1** [comment]
Closing this bug.  I ended up resolving my specific issue by adjusting the voltage that the VFD associates with 100% speed.  In my case, the Fuling DZB312B005.5L2DK VFD needed parameter F2.11 to be reduced from 10.00 to 9.58 to work correctly with the cncdrive AXBB-E controller along with a 1K inline resistor.

Now requested speed vs. measured speed is +/-50 RPM from 2000-24000 RPM.

Thanks for all your help @modbog.
